### PR TITLE
*: add a procedure that upgrades data from v2 to v3

### PR DESCRIPTION
**Pull request by overvenus**
## What have you changed? (mandatory)

In v3.x, we move all meta data to raft engine and remove RAFT CF of kv engine. This is a precede change of #4372 , it add a procedure that upgrades data from v2 to v3:

- Moves store_ident_key/prepare_boostrap_key from kv.default to raft.default
- Moves apply_state_key/region_state_key from kv.raft to raft.default
- Remove **snapshot_raft_state_key**

## What are the type of the changes? (mandatory)

- Engineering (engineering change which doesn't change any feature or fix any issue)

## How has this PR been tested? (mandatory)

Unit tests and integration tests, however integration tests can not run with out full change of #4372 .

## Does this PR affect documentation (docs) update? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

## Refer to a related PR or issue link (optional)

#4372

### CODE REVIEW

**Comment by 5kbpers** [comment]
LGTM

File: src/raftstore/store/peer_storage.rs [hunk 1]
```diff
@@ -1470,6 +1471,166 @@ impl Storage for PeerStorage {
...
+    kv_db_opts: DBOptions,
+    kv_cfg: &config::DbConfig,
+) {
+    use crate::storage::CF_RAFT;
+    if !util::rocksdb_util::db_exist(kv_path) {
+        debug!("no need upgrade to v3.x");
+        return;
+    }
+
+    if DB::list_column_families(&kv_db_opts, kv_path)
```
Comment (BusyJay) [inline src/raftstore/store/peer_storage.rs]: What happen if the number of cfs in `kv_db_opts` doesn't match the actual db cfs?

Comment (overvenus) [reply src/raftstore/store/peer_storage.rs]: `DBOptions` does not contain cf.

File: src/raftstore/store/keys.rs [hunk 2]
```diff
@@ -146,29 +151,34 @@ fn make_region_meta_key(region_id: u64, suffix: u8) -> [u8; 11] {
     key
 }
 
-// Decode region meta key, return the region key and meta suffix type.
-pub fn decode_region_meta_key(key: &[u8]) -> Result<(u64, u8)> {
-    if REGION_META_PREFIX_KEY.len() + mem::size_of::<u64>() + mem::size_of::<u8>() != key.len() {
+/// Decode region category key, return the region id and meta suffix type.
+fn decode_region_category_key(prefix: &[u8], key: &[u8], which: &str) -> Result<(u64, u8)> {
```
Comment (BusyJay) [inline src/raftstore/store/keys.rs]: Oh, what I mean is `which` is meaningless without context.

File: src/raftstore/store/peer_storage.rs [hunk 3]
```diff
@@ -1470,6 +1471,150 @@ impl Storage for PeerStorage {
...
+                        "snapshot_raft_state" => ?snapshot_raft_state,
+                        "raft_state" => ?raft_state
+                    );
+                }
+                return Ok(true);
+            }
+        } else if let Ok((region_id, suffix)) = keys::decode_region_meta_key(key) {
+            if suffix == keys::REGION_STATE_SUFFIX {
+                upgrade_raft_wb.put(key, value)?;
+                info!("upgrading region state"; "region_id" => region_id, "value" => ?value);
```
Comment (BusyJay) [inline src/raftstore/store/peer_storage.rs]: Seems too verbose.

**Pull request merged**